### PR TITLE
Try to make ObligationForest more efficient

### PR DESCRIPTION
**Pull request by bugadani**
This PR tries to decrease the number of allocations in ObligationForest, as well as moves some cold path code to an uninlined function.

### CODE REVIEW

**Comment by petrochenkov** [comment]
Didn't notice the failing CI, the try build is going to fail.
Needs green CI to run perf.

**Comment by bugadani** [comment]
The CI fails because the tests don't compile, I don't know if perf requires that. Working on it :)

**Comment by bugadani** [comment]
@petrochenkov should be good now.

**Comment by bugadani** [comment]
Can I request another perf run, please?

**Comment by petrochenkov** [comment]
@bors try @rust-timer queue

FYI, a lot of effort went into benchmarking and performance-tuning this code in the past, see the git history for obligation forest, commits by nnethercote in particular.
So it's quite possible that something like this PR was already tried and rejected.

**Comment by bugadani** [comment]
@petrochenkov thanks! I'm not trying to claim that I can do better in a day, I don't plan on pushing this further if the results aren't good enough. But seeing the results can be an educational experience for sure.

**Comment by panstromek** [comment]
Just in case you didn't notice already, there's also another big PR open to improve perf for Obligation Forest. https://github.com/rust-lang/rust/pull/69218

**Comment by bugadani** [comment]
Wow that's big, but looks like a work in progress. I'm interested in the "death by thousand cuts" types of improvements that are both somewhat glaring and less risky.

**Comment by bugadani** [comment]
That looks better 😅

**Comment by petrochenkov** [comment]
I'm not familiar with obligation forest, but people who are familiar are mostly unavailable, so I'll try to review this during the weekend, the changes don't look too invasive.
cc @nnethercote just in case

**Comment by bugadani** [comment]
@petrochenkov Thanks!

With the last force push I've removed a commit-revert pair that doesn't need to be there.

File: compiler/rustc_data_structures/src/obligation_forest/mod.rs [hunk 1]
```diff
@@ -251,12 +251,21 @@ enum NodeState {
     Error,
 }
 
+/// This trait allows us to make Outcome an opaque type, allowing a different implementation
+/// for tests.
```
Comment (nnethercote) [inline compiler/rustc_data_structures/src/obligation_forest/mod.rs:255]: I didn't understand this comment at first. Maybe change it to this: "This trait allows us to have two implementations of `Outcome`: the normal one, and one that does some additional work and checking for use in tests."

**Review by nnethercote — CHANGES_REQUESTED** [review]
These look ok to me. This will interfere with #69218, but that PR is in a very uncertain state so I think it's reasonable to go forward with these concrete improvements.

@bugadani: How did you find these improvements? Did you use a profiler, and if so, which one? I'm alway curious to learn how other people find optimizations :)

File: compiler/rustc_data_structures/src/obligation_forest/mod.rs (hunk 1 above)
Comment (bugadani) [reply compiler/rustc_data_structures/src/obligation_forest/mod.rs]: Thanks, I've updated this.

**Comment by bugadani** [comment]
@nnethercote in this case this was complete guesswork. I was looking a piece of related code somewhere, wanted to understand what it does, and stumbled upon the comments in OF mentioning the code is hot. From there, I went on looking if the implementation does work it doesn't necessarily have to.

There still are some unnecessary `iterator -> vector -> iterator` flows around `process_obligation`, which unnecessarily copy their data to the heap. But this one is harder to avoid.

**Comment by pietroalbini** [comment]
@bors treeclosed=1000

Need to fix another thing in the bors setup.

**Comment by Mark-Simulacrum** [comment]
Final [results](https://perf.rust-lang.org/compare.html?start=cb2462c53f2cc3f140c0f1ea0976261cab968a34&end=f90e6173053f7e6b377d7f75367b511ceee7d9d1&stat=instructions:u) are as expected.

**Pull request merged**